### PR TITLE
Restrict domain of number theoretic functions to N not Z.

### PR DESCRIPTION
**Pull request by wbhart**
After consulting many references, I discovered that multiplicative number theoretic functions can either be defined as function on N or functions on Z. However, I can find no source documenting a consistent set of definitions for Z for all standard number theoretic functions. In fact, *every single source* I consulted defines them over N, even if they say they are defining them over Z, and then goes to extreme lengths to ensure they only use the definition over N, e.g. by inserting absolute values every time the function is used.

The problem is particularly severe for Euler's phi function, where phi(0) could be either undefined, 0, 1 or 2, depending on the definition used. And at negative values, there are two common conventions, depending again on the definition.

I strongly suggest that people define the function phi (instead of euler_phi) if they do not mean the multiplicative arithmetic function defined on N.

Similarly, the divisor_sigma function is not defined at 0, since zero has infinitely many divisors. Again, I suggest defining the function sigma (or something else of one's own choosing) if another definition is expected.

For this reason, the current pull request restricts *all* multiplicative arithmetic functions to positive inputs. 

*If* there was a reference that dealt with consistent definitions over Z for all the multiplicative arithmetic functions, and *if* this was widely used, I would/could take a different opinion on this matter. But as it is, it is very clear that a lot of people will be pissed off by any definition we give for these functions at 0 or negative values.

### CODE REVIEW

**Comment by wbhart** [comment]
I realise that restricting divisor_sigma to positive values only is somewhat artificial. After all, one can easily take the definition that it sums the n-th powers of the positive divisors of m, which is defined for all m except zero. However, the justification for also restricting this function is that if you do not, someone will use its definition at m < 0 as a justification for using their personal preferred definition of one of the other functions for m < 0.

This is particularly acute for phi(m) where you can either take the (by far more common) definition:

"The number of positive integers x in the range 1 <= x <= m for which gcd(m, x) = 1"

or one can use the fact that phi(m) is equal to |(Z/mZ)*| for positive m and take that as the definition (not that I can find a single standard textbook that actually does this).

It is far better to avoid the disagreements altogether and not have an opinion on the definition of these functions at values < 1.

The standard fact about multiplicative arithmetic functions on N is that they form a group under Dirichlet convolution. This clearly supports the idea that the functions are defined on N, not Z, as only the values in N are needed in the definition. It is incredibly hard to see what the definitions should be if this is extended to functions on Z.

**Comment by fredrik-johansson** [comment]
The Fibonacci numbers are not like the others.

**Comment by wbhart** [comment]
Yeah, I will change Fibonaccci to be defined everywhere.

File: src/flint/fmpz.jl [hunk 1]
```diff
@@ -1304,29 +1304,27 @@ end
 @doc Markdown.doc"""
     fibonacci(x::Int)
 >  Return the $x$-th Fibonacci number $F_x$. We define $F_1 = 1$, $F_2 = 1$ and
-> $F_{i + 1} = F_i + F_{i - 1}$ for all $i > 2$. We require $n \geq 0$. For
-> convenience, we define $F_0 = 0$.
+> $F_{i + 1} = F_i + F_{i - 1}$ for all $i > 2$. The sequence is defined for
+> all integers.
```
Comment (rfourquet) [inline src/flint/fmpz.jl]: This begs the question, how is the sequence defined for `i < 0`. I don't know the best way to formulate it, but maybe just replace `for all $i > 2$.` by `for all $i$` ? I can make a PR to update that if you like.

Comment (wbhart) [reply src/flint/fmpz.jl]: Yeah, you can change it to "for all i" if you wish.

Comment (rfourquet) [reply src/flint/fmpz.jl]: done: #733

**Pull request merged**